lightwood/encoder/datetime/datetime.py

```diff
@@ -1,90 +1,82 @@
-import datetime
-import calendar
-from typing import Optional
+from typing import Union
+
 import torch
+import numpy as np
+import pandas as pd
+
 from lightwood.encoder.base import BaseEncoder
-from lightwood.helpers.general import is_none
+from lightwood.helpers.log import log
 
 
 class DatetimeEncoder(BaseEncoder):
     """
     This encoder produces an encoded representation for timestamps.
 
-    The approach consists on decomposing the timestamp objects into its constituent units (e.g. day-of-week, month, year, etc), and describing each of those with a single value that represents the magnitude in a sensible cycle length.
+    The approach consists on decomposing the timestamp objects into its constituent units (e.g. month, year, etc), and describing each of those with a single value that represents the magnitude in a sensible cycle length.
     """  # noqa
     def __init__(self, is_target: bool = False):
         super().__init__(is_target)
-        self.fields = ['year', 'month', 'day', 'weekday', 'hour', 'minute', 'second']
-        self.constants = {'year': 3000.0, 'month': 12.0, 'weekday': 7.0,
-                          'hour': 24.0, 'minute': 60.0, 'second': 60.0}
-        self.output_size = 7
+        self.constant_keys = ['year', 'month', 'day', 'hour', 'minute', 'second']
+        self.constant_vals = torch.Tensor([3000.0, 12.0, 31.0, 24.0, 60.0, 60.0])  # cycle length
+        self.constant_map = {k: v.item() for k, v in zip(self.constant_keys, self.constant_vals)}
 
-    def prepare(self, priming_data):
-        if self.is_prepared:
-            raise Exception('You can only call "prepare" once for a given encoder.')
+        self.output_size = len(self.constant_keys)
+        self.empty_vector = np.zeros((self.output_size, ))
 
-        self.is_prepared = True
+        self.max_vals = torch.Tensor([pd.Timestamp.max.year - 1, 12, 31, 23, 59, 59])
+        self.min_vals = torch.Tensor([pd.Timestamp.min.year + 1, 1, 1, 0, 0, 0])
 
-    def encode(self, data):
-        """
-        :param data: # @TODO: receive a consistent data type here; currently either list of lists or pd.Series w/lists
-        :return: encoded data
-        """
-        if not self.is_prepared:
-            raise Exception('You need to call "prepare" before calling "encode" or "decode".')
-
-        ret = [self.encode_one(unix_timestamp) for unix_timestamp in data]
-
-        return torch.Tensor(ret)
+    def prepare(self, priming_data):
+        self.is_prepared = True
 
-    def encode_one(self, unix_timestamp: Optional[float]):
+    def encode(self, data: Union[np.ndarray, pd.Series]) -> torch.Tensor:
```

Comment:
I like what's being done here, very nice!

My only concern is that that I can see the output tensor being somewhat discontinuous: for instance, when the weekday changes, you will suddenly go from an entry with 'weekday = 6' to 'weekday = 0'; I am not entirely sure this will help the ML side of things (more inclined to think it will not).

Something that worked for me is to use trigonometric encoders for the time. Basically, every component (year, month, day, weekday, hour, mine, second) is encoded with a sine and a cosine function with a period equal to the component's period (365 days per year, 7 days a week, 24 hours a day, etc). This not only puts the numbers in the [-1, 1] range (very nice for ML) but it also avoids nasty discontinuity artifacts, like changing from weekday 6 to 0 from one step to the next one. 

Might be worth giving this a thought if the encoder is not performing as expected ;-)

Reply:
Yes, I totally agree, we have the `datetime_sin_encoder.py` for this, albeit not as a default option. It's next in line to be vectorized, so more on that later. As for improvements w.r.t. this one, I don't think we've actually ran the benchmark suite on it, would be a nice next step!